src/frontend/src/optimizer/plan_node/logical_source_backfill.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ops::Bound;
+use std::rc::Rc;
+
+use pretty_xmlish::{Pretty, XmlNode};
+use risingwave_common::util::iter_util::ZipEqFast;
+use risingwave_connector::parser::additional_columns::add_partition_offset_cols;
+
+use super::generic::{GenericPlanRef, SourceNodeKind};
+use super::stream_watermark_filter::StreamWatermarkFilter;
+use super::utils::{childless_record, Distill};
+use super::{
+    generic, BatchProject, BatchSource, ColPrunable, ExprRewritable, Logical, LogicalFilter,
+    LogicalProject, LogicalSource, PlanBase, PlanRef, PredicatePushdown, StreamProject,
+    StreamRowIdGen, ToBatch, ToStream,
+};
+use crate::catalog::source_catalog::SourceCatalog;
+use crate::error::Result;
+use crate::expr::{ExprImpl, ExprRewriter, ExprVisitor};
+use crate::optimizer::optimizer_context::OptimizerContextRef;
+use crate::optimizer::plan_node::expr_visitable::ExprVisitable;
+use crate::optimizer::plan_node::utils::column_names_pretty;
+use crate::optimizer::plan_node::{
+    ColumnPruningContext, PredicatePushdownContext, RewriteStreamContext, StreamSourceBackfill,
+    ToStreamContext,
+};
+use crate::optimizer::property::Distribution::HashShard;
+use crate::utils::{ColIndexMapping, Condition};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct LogicalSourceBackfill {
+    pub base: PlanBase<Logical>,
+    pub core: generic::Source,
+
+    /// Expressions to output. This field presents and will be turned to a `Project` when
+    /// converting to a physical plan, only if there are generated columns.
+    output_exprs: Option<Vec<ExprImpl>>,
+    /// When there are generated columns, the `StreamRowIdGen`'s row_id_index is different from
+    /// the one in `core`. So we store the one in `output_exprs` here.
+    output_row_id_index: Option<usize>,
+}
+
+impl LogicalSourceBackfill {
+    pub fn new(source_catalog: Rc<SourceCatalog>, ctx: OptimizerContextRef) -> Result<Self> {
+        let mut column_catalog = source_catalog.columns.clone();
+        let row_id_index = source_catalog.row_id_index;
+
+        let kafka_timestamp_range = (Bound::Unbounded, Bound::Unbounded);
```

Comment:
Hard-coded?

Reply:
This is copy-pasted code from LogicalSource. And the kafka timestamp stuff does need refinement. https://github.com/risingwavelabs/risingwave/issues/14636 https://github.com/risingwavelabs/risingwave/issues/14627